src/test/crimson/test_messenger_recovery.cc

```diff
@@ -0,0 +1,801 @@
+         std::pair<crimson::net::MessengerRef, crimson::net::MessengerRef> &p =
+           available_connections[conn];
+         if (!p.first->get_default_policy().server &&
+             !p.second->get_default_policy().server) {
+             //verify that equal-to operator applies here
+        ceph_assert(conn->get_messenger() == p.first.get());
+        crimson::net::ConnectionRef peer = p.second->connect(
+          p.first->get_myaddr(), p.first->get_mytype());
+        peer->mark_down();
+        dispatcher.clear_pending(peer);
+        available_connections.erase(peer);
+         }
+       }
+       ceph_assert(available_connections.erase(conn) == 1U);
+       return seastar::now();
+     });
+   }
+
+   void print_internal_state(bool detail=false) {
+     logger().info("available_connections: {} inflight messages: {}",
+       available_connections.size(),
+       dispatcher.get_pending());
+     if (detail && !available_connections.empty()) {
+       dispatcher.print();
+     }
+   }
+
+   seastar::future<> wait_for_done() {
+     int i = 0;
+     return seastar::do_until([this] {
+       return !dispatcher.get_pending(); },
+     [this, &i] {
+       if (i++ % 50 == 0){
+         print_internal_state(true);
+       }
+       return seastar::sleep(100ms);
+     }).then([this] {
+       return seastar::do_for_each(available_servers, [] (auto server) {
+         logger().warn("server {} shutdown" , server->get_myaddrs());
+         server->stop();
+         return server->shutdown();
+       });
+     }).then([this] {
+       available_servers.clear();
+     }).then([this] {
+       return seastar::do_for_each(available_clients, [] (auto client) {
+         logger().warn("client {} shutdown" , client->get_myaddrs());
+         client->stop();
+         return client->shutdown();
+       });
+     }).then([this] {
+       available_clients.clear();
+     });
+   }
+
+   void handle_reset(crimson::net::ConnectionRef con) {
+     available_connections.erase(con);
+     dispatcher.clear_pending(con);
+   }
+
+   int servers_count() {
+     return available_servers.size();
+   }
+
+   int clients_count() {
+     return available_clients.size();
+   }
+
+   int connections_count() {
+     return available_connections.size();
+   }
+};
+
+void SyntheticDispatcher::ms_handle_reset(crimson::net::ConnectionRef con, bool is_replace) {
+    workload->handle_reset(con);
+  }
+
+// Testing Crimson messenger (with msgr-v2 protocol) robustness against
+// network delays and failures. The test includes stress tests and
+// socket level delays/failures injection tests, letting time
+// and randomness achieve the best test coverage.
+
+// Test Parameters:
+// Clients: 8             (stateful)
+// Servers: 32            (lossless)
+// Connections: 100       (Generated between random clients/server)
+// Random Operations: 500 (Generate/Drop Connection, Send Message, Sleep)
+seastar::future<> test_stress_1()
```

Comment:
IIUC test_stress_1 and 2 are only different from the parameters about intensity, not about the scenario. So why not just keep only one of them. Same applies to injection_1 and 2.

Reply:
The original idea was also to test different policies combinations. Although I believe this test is exhaustive as is (Including the validation of `loseless_peer`), so I will keep only one of them for now.